### PR TITLE
Stripe enhancements

### PR DESCRIPTION
**Pull request by CompuIves**
- Support for reactivating subscriptions
- Support for coupon codes

### CODE REVIEW

File: packages/app/src/app/components/SubscribeForm/CheckoutForm/index.tsx [hunk 1]
```diff
@@ -96,6 +133,19 @@ class CheckoutForm extends React.PureComponent {
           />
         </CardContainer>
 
+        {!noCoupon && (
```
[the remark this answers is not in the document]
Comment (CompuIves) [reply packages/app/src/app/components/SubscribeForm/CheckoutForm/index.tsx]: Agree! Let's do that!

File: packages/app/src/app/pages/Patron/PricingModal/PricingChoice/ChangeSubscription/index.js [hunk 2]
```diff
@@ -23,16 +34,36 @@ function ChangeSubscription({ date, store, signals }) {
...
+          Update
+        </StyledButton>
+      </Buttons>
+    </>
   );
 
+  if (markedAsCancelled) {
+    buttons = (
+      <Buttons>
+        <StyledButton onClick={() => updateSubscription({ coupon: '' })}>
```
[the remark this answers is not in the document]
Comment (CompuIves) [reply packages/app/src/app/pages/Patron/PricingModal/PricingChoice/ChangeSubscription/index.js]: Yep, that's allowed. But this scenario is only for reactivating a cancelled subscription. In the other cases you can apply a coupon.

File: packages/app/src/app/pages/common/Modals/PreferencesModal/PaymentInfo/index.js [hunk 3]
```diff
@@ -35,6 +35,7 @@ class PaymentInfo extends React.Component {
         <SubscribeForm
           buttonName="Update"
           loadingText="Updating Card Info..."
+          noCoupon
```
[the remark this answers is not in the document]
Comment (CompuIves) [reply packages/app/src/app/pages/common/Modals/PreferencesModal/PaymentInfo/index.js]: No, they're enabled. But I didn't want the coupon input to show for the place where people can update their payment details. Only wanted to show it on /patron. This component is used in both places.

File: packages/app/src/app/components/LinkButton/LinkButton.tsx [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  background-color: transparent;
+  border: none;
+  outline: none;
+
+  color: ${props => props.theme.secondary};
+  text-decoration: underline;
+  cursor: pointer;
+
+  &:hover {
+    color: ${props => props.theme.secondary.lighten(0.1)};
```
[the remark this answers is not in the document]
Comment (CompuIves) [reply packages/app/src/app/components/LinkButton/LinkButton.tsx]: Hmm, I prefer not to destructure at the top. It's less readable for me (which properties are dynamic?) and whenever we introduce an interpolated variable in an existing component without variables we would need to wrap everything in a `css`, which generates a bigger diff than necessary.

**Comment by CompuIves** [comment]
Good points! I'll make sure to merge https://github.com/codesandbox/codesandbox-client/commit/43942f9085050836f38c15c71e8c198c40325e48 in this PR and fix the mentioned points.

**Comment by Saeris** [comment]
I can do another pass to convert all touched files to TypeScript and add types in two additional commits, or we can save that for another PR. Besides that the only other things I think we should do are:

- convert any class components to functional components with hooks
- update import/export statements to use named exports

File: packages/app/src/app/pages/Patron/PricingModal/PricingChoice/index.js [hunk 5]
```diff
@@ -21,63 +20,71 @@ import {
...
+      priceChanged,
+      createSubscriptionClicked,
+      updateSubscriptionClicked,
+      cancelSubscriptionClicked,
+    },
+  } = useSignals();
+  const {
+    isLoggedIn,
+    isPatron,
+    user: { name, subscription },
```
Comment (CompuIves) [inline packages/app/src/app/pages/Patron/PricingModal/PricingChoice/index.js]: I prefer to do less destructuring here, currently it:

1. Removes the context of the property (`name` of what?)
2. Opens up the chance of bugs (what if `user` is undefined and only rendered under a condition?)

I think we should keep deep destructuring to a minimum, we can discuss it shortly later though. It's a quite small thing, so preferably we won't put much time into discussing it.

**Pull request merged**